### ISSUE TITLE
MathJax upgrade to v2.5.x

### ISSUE THREAD

**Issue by danielweck** [readium/readium-js-viewer#262]
See:
http://www.mathjax.org/mathjax-v2-5-beta-now-available/

Readium makes use of a pre-packaged single-file MathJax library (SVG rendering, no external fonts dependency):
https://github.com/readium/readium-js-viewer/blob/develop/lib/mathjax/MathJax.js
...which can be rebuilt using this Grunt build script:
https://github.com/pkra/MathJax-single-file

**Comment by danielweck** [comment]
No need to replicate this issue for each repository (native application "launcher"), so here are direct links to bundled MathJax libs:

Android:
https://github.com/readium/SDKLauncher-Android/blob/develop/SDKLauncher-Android/assets/reader-payloads/MathJax.js

OSX:
https://github.com/readium/SDKLauncher-OSX/blob/develop/LauncherOSX/ReaderScripts/mathjax/MathJax.js

iOS:
https://github.com/readium/SDKLauncher-iOS/blob/develop/Resources/mathjax/MathJax.js

Windows (not there yet):
https://github.com/readium/SDKLauncher-Windows/tree/develop

**Comment by danielweck** [comment]
CC @pkra

**Comment by pkra** [comment]
Thanks for pinging me. We might want to discuss how to build MathJax for Readium. There are a couple of new options on the table now.

**Comment by pkra** [comment]
But that can wait until 2.5 is released.

**Comment by pkra** [comment]
As 2.5 is getting close to the release, I've done a quick test run with the single-file builder; a few bugs popped up but nothing too serious. 

The question is if the setup of MathJax in Readium should be changed. There was some feedback on the dev-list a while back. It comes down to which components to ship, in particular which output to use/ship. Not sure where to discuss this / reach the most people.

**Comment by danielweck** [comment]
Thanks @pkra 
I believe that the rationale for using a single-file SVG-renderer build of MathJax in Readium, is to:

- reduce the library footprint
- simplify the source tree (I remember seeing many, many JS files in the early builds)
- avoid external dependencies such as fonts

...anyone to confirm?

**Comment by pkra** [comment]
Thanks, @danielweck. The first two points are independent of what I had in mind. The last one is relevant if additional output options are integrated in which case some fonts are needed. I think there were some complaints on SVG output in general (too thick etc). There are certain advantages for the other outputs in terms of CSS inheritance (which came up also).

Anyway, I'd be happy to discuss this and tweak the single-file build accordingly. But obviously the build can also stay as is.

**Comment by danielweck** [comment]
Some thoughts:

- Currently, MathJax is bundled with all Readium flavours (i.e. Chrome extension, cloud reader, native application launchers), and this functionality cannot be deactivated by *users* (only by developers / integrators / vendors). Perhaps we should consider making this an opt-out feature, via user preferences? Possible use-cases:

-- As the browser-oriented flavour of Readium naturally lends itself to online APIs / services (e.g. audible speech, dictionary, annotations, etc.), the original MathML markup can be passed as input to an external tool like ; for the sake of argument ; Benetech's MathML Cloud which offers alternative representations / renditions.

-- Some web browsers / engines may provide "native" or alternative support for MathML (?)

- The MathJax library is currently built specifically for Readium usage, but only SVG output is enabled. Is it technically possible to build a more versatile version of MathJax, which ideally could be configured programmatically (at runtime) so that users would be able to dynamically try various rendering options via the regular Readium control panel / preference pane?

- Right now, Readium injects the single-file mathjax.js library into each loading HTML document, but this JS could just be some sort of bootstrapper for a more complex set of files / dependencies. My point is: should we decide to change the way MathJax is used in Readium, I would hope that the refactoring effort in browser and native apps to be minimal.

**Comment by pkra** [comment]
>  Perhaps we should consider making this an opt-out feature, via user preferences? 

Sounds great. I've noticed a couple of Readium-based apps that do not use MathJax in WebKit environments -- which sucks for users given that WebKit's MathML support is not anywhere close to Gecko or MathJax.

>  Benetech's MathML Cloud which offers alternative representations / renditions.

Just FYI, MathML Cloud uses tools developed by MathJax and we will pull similar features into core MathJax this year.

> Some web browsers / engines may provide "native" or alternative support for MathML (?)

Right. See my comments above.

> Is it technically possible to build a more versatile version of MathJax, which ideally could be configured programmatically (at runtime) so that users would be able to dynamically try various rendering options via the regular Readium control panel / preference pane? 

I had developers (not users) in mind but this is what I was suggesting. 

If the single-file build includes more MathJax components, developers (or even users) would have the option of cherry picking. For example, MathJax provides a "Native MathML" output, which tweaks MathML to work around bugs and missing MathML features in Gecko and WebKit. A developer might want to activate that mode as a compromise between speed and layout quality -- get features `mlabeledtr` or elementary math from MathJax but rendering at native speed.

**Comment by pkra** [comment]
Addendum to the last sentence: in particular, devs could switch modes in response to the user platform. E.g., on Gecko they might be happy without MathJax, but on WebKit they might want the tweaks from MathJax, on mobile Chrome and IE they might want to use the new CommonHTML output etc etc..

**Comment by pkra** [comment]
FYI 2.5 was released last week. 

The existing build should be modified to include`mml3.js` instead of `content-mathml.js` (unless you want to keep Content MathML support which is not required by epub3). The new fast-preview mode might also be interesting; it uses the new, very fast (but incomplete) CommonHTML output as a preview.

Should I open a thread on the mailing list?

**Comment by danielweck** [comment]
Please do @pkra ("open a thread on the mailing list").
If Content MathML support is not mandatory in EPUB3, and assuming there are potential performance gains in skipping this MathJax feature (fewer bytes, less rendering overhead?), then it makes sense to update the Readium build accordingly.
As for the CommonHTML output: what does "fast preview" mean? (would that be a rendering stage prior to the full SVG output, similar to what can be achieved using the image specified in the `altimg` attribute?)

**Comment by pkra** [comment]
> As for the CommonHTML output: what does "fast preview" mean?  (would that be a rendering stage prior to the full SVG output, similar to what can be achieved using the image specified in the altimg attribute?)

Prettty much. The fast preview is ~10x faster but not yet on the level of other outputs. See this example http://cdn.mathjax.org/mathjax/latest/test/sample-eqnum.html for how it looks. This output will eventually reach the same quality as the others while retaining most of its speed.

**Comment by pkra** [comment]
Since there's been no further interest, I've updated https://github.com/pkra/MathJax-single-file as I saw fit, see https://github.com/pkra/MathJax-single-file/commit/3d58074f614699c496854f540b5c86d75df74706

**Comment by danielweck** [comment]
Thanks @pkra much appreciated!
For MathML-interested Readium folks who missed Peter Krautzberger's email on the readium-dev Google group (request for feedback): https://groups.google.com/forum/#!topic/readium-dev/sTXjx6GqobA

Looking at your README, is that the same configuration as the previous Readium-MathJax build?

- MathML input
- SVG output (with  blacker:1 )
- Mathjax TeX "fonts"
- extensions: mml3.js, webfonts matching
- misc. MathJax internals


Just a note about the Grunt task `shell:prepTempHelpers:command` (there's a `TODO` comment for this "hack"): I guess this will eventually be replaced with a more portable technique (the script currently works on sh/Cygwin, but not MS-DOS/Powershell)
See: https://github.com/pkra/MathJax-single-file/blob/3d58074f614699c496854f540b5c86d75df74706/Gruntfile.js#L41

**Comment by pkra** [comment]
> Looking at your README, is that the same configuration as the previous Readium-MathJax build?

Not quite but close enough. I dropped the ContentMathML extension in favor of the mml3 extension. The old ContentMathML extension was two hacks combined; it's been split into two and mml3 has the relevant stuff for Readium (for elementary math support); ContentMathML is not included in epub3 but could be readded whenver you want.

> (the script currently works on sh/Cygwin, but not MS-DOS/Powershell)

Always happy to get a pull request but, yes, that's a hack.

**Comment by danielweck** [comment]
Thank you so much Peter for taking the time!
I don't have spare cycles at the moment to study and contribute to the build script, so I hope other MathML-interested devs can chime in to help.

**Comment by danielweck** [comment]
Readium cloud-reader and Chrome extension/packaged-app now fetch the latest (v2.5.1) MathJax single-file build directly from Peter Krautzberger's GitHub repository https://github.com/pkra/MathJax-single-file via NPM dependency (don't forget to invoke the build task `npm run prepare`).

Code checkin: https://github.com/readium/readium-js-viewer/commit/ab88cb281455d416e0c1406db687ec70e007879f

TODO: native launcher apps, see original comment above:
https://github.com/readium/readium-js-viewer/issues/262#issuecomment-69061606

**Comment by danielweck** [comment]
@rkwright I created a new `MathJax` label:

https://github.com/readium/readium-shared-js/labels/MathJax
https://github.com/readium/readium-js-viewer/labels/MathJax

**Comment by danielweck** [comment]
Waffle MathJax filter: https://waffle.io/readium/readium-js-viewer?label=MathJax

**Comment by pkra** [comment]
[This](https://waffle.io/readium/readium-js-viewer/cards/544821477abdfe69dab9a51f) should be fixable in the upcoming release via a new extension we're building; not ideal but what AT vendors suggested.

**Comment by pkra** [comment]
At this point, I would wait until 2.6 (hitting beta next week *fingers crossed*). 

2.6 brings various accessibility improvements. It also fixes that menu image error issue. There's also a new output you might want to consider (faster, HTML-based).

**Comment by danielweck** [comment]
@pkra Great! 
Because of testing costs, we prefer to reduce the matrix of possibilities, so we will need to choose either the SVG output mode, or the faster HTML mode (at least, in terms of what cloud-reader and Chrome app flavours we deploy in the wild). We also have the imperative of single-file MathJax build. How does the HTML output deal with fonts / glyphs? Are there other external dependencies?

**Comment by pkra** [comment]
> How does the HTML output deal with fonts / glyphs? 
> Are there other external dependencies?

It uses fonts, yes. Might try data-uri-ying them or hard-coding an unusual path; we'll see. The biggest advantage might be that you can cache the output -- the generated markup is identical across all supported browsers (IE9+, just like the SVG one). Not sure Readium is set up to optimize that kind of thing though.

**Comment by danielweck** [comment]
Broken build:
https://github.com/readium/readium-js-viewer/issues/475

**Comment by danielweck** [comment]
MathJax 2.6 now integrated into develop branch:

https://readium.firebaseapp.com/?epub=https%3A%2F%2Fcdn.rawgit.com%2FIDPF%2Fepub-testsuite%2Fmaster%2Fcontent%2F30%2Fepub30-test-0100&goto=%7B%22idref%22%3A%22xhtml-mathml-001%22%7D

**Comment by pkra** [comment]
The last bit of conversation was about trying the new output. I'll probably play around with that at some point next month for another project. But if there's interest from Readium, please let me know.
